lib/base/atomic.hpp

```diff
@@ -41,6 +41,35 @@ class Atomic : public std::atomic<T> {
 	}
 };
 
+/**
+ * Classic spinlock mutex based on std::atomic_flag.
+ *
+ * @ingroup base
+ */
+class SpinMutex
+{
+public:
+	SpinMutex() = default;
+	SpinMutex(const SpinMutex&) = delete;
+	SpinMutex(SpinMutex&&) = delete;
+	SpinMutex& operator=(const SpinMutex&) = delete;
+	SpinMutex& operator=(SpinMutex&&) = delete;
+
+	void lock() noexcept
+	{
+		while (m_State.test_and_set(std::memory_order_acquire)) {
+		}
+	}
+
+	void unlock() noexcept
+	{
+		m_State.clear(std::memory_order_release);
+	}
+
+private:
+	std::atomic_flag m_State = ATOMIC_FLAG_INIT;
+};
```

Comment:
Why do you think a spinlock is a good idea here?

Let me quote [cppreference.com for `std::atomic_flag`](https://en.cppreference.com/w/cpp/atomic/atomic_flag):

> A [spinlock](https://en.wikipedia.org/wiki/Spinlock) mutex demo can be implemented in userspace using an atomic_flag. Do note that spinlock mutexes are [extremely dubious](https://www.realworldtech.com/forum/?threadid=189711&curpostid=189723) in practice.

And we even had problems in Icinga 2 in the past due to a spinlock, see #8507/#8488. So I have doubts that would be a good idea this time.

Reply:
> Why do you think a spinlock is a good idea here?

std::mutex is 5 pointers large on x64, whereas std::atomic_flag at most one.

> > spinlock mutexes are [extremely dubious](https://www.realworldtech.com/forum/?threadid=189711&curpostid=189723) in practice.

But the lock scope here is very small, isn't it? Not to mention how unlikely you'll hit the slow path at all in our case...

Reply:
> But the lock scope here is very small, isn't it?

Depends on what's inside. If it's an intrusive pointer, it just needs to increment the reference count, if it's a string though, it has to copy it, including allocating memory for it.

> Not to mention how unlikely you'll hit the slow path at all in our case...

How did you conclude it's unlikely? #9364 was done for a reason, there were random crashes because. Crash means concurrent access, so that will go into the slow path. Also note that not every concurrent access resulted in a crash, so there will be more "hitting the slow path" than there were crashes.

Reply:
Is [this change](https://github.com/Icinga/icinga2/compare/b7611ed134d3c4ad2a3683a554fff96445ad2e49..37b4d107d9a29ddbe723cc239114d1b8123ce3e6#diff-c8f9afd77a5aca1303178c2b529c50b9fd1a62c7b5fba5e92110a472ad9466a3) hidden in a big rebase push related to this conversation? If so, please comment why you think this resolves it. Otherwise, what was the motivation for that change and how will you address this?

Reply:
> > But the lock scope here is very small, isn't it?
> 
> Depends on what's inside. If it's an intrusive pointer, it just needs to increment the reference count, if it's a string though, it has to copy it, including allocating memory for it.

You're right. Especially the latter case can take a few more CPU cycles. So instead of wasting them on pure spinning in a waiting thread, the latter now says ["after you"](https://en.cppreference.com/w/cpp/thread/yield) while something's locked. This shall bring the mutex holder in foreground and let it finish and unlock.